From dfc1ac0ac6eb33549bc3846181cdc8e70509f3af Mon Sep 17 00:00:00 2001
From: Yongbo Zhang <yongbo.zhang.xh@renesas.com>
Date: Thu, 16 Dec 2021 17:56:57 +0800
Subject: [PATCH 16/27] mfis: irqbalance: set the initial affinity to prevent
 every interrupt being on core0

before:
console:/ # cat /proc/interrupts | grep mfis
14:      38349          0          0          0          0          0          0          0     GIC-0 256 Level     e6260000.mfis
15:        872          0          0          0          0          0          0          0     GIC-0 257 Level     e6260000.mfis
16:         15          0          0          0          0          0          0          0     GIC-0 258 Level     e6260000.mfis
17:          1          0          0          0          0          0          0          0     GIC-0 259 Level     e6260000.mfis
18:          0          0          0          0          0          0          0          0     GIC-0 260 Level     e6260000.mfis
19:          0          0          0          0          0          0          0          0     GIC-0 261 Level     e6260000.mfis
20:          0          0          0          0          0          0          0          0     GIC-0 262 Level     e6260000.mfis
21:          0          0          0          0          0          0          0          0     GIC-0 263 Level     e6260000.mfis

now:
console:/ # cat /proc/interrupts | grep mfis
14:          0     578833          0          0          0          0          0          0     GIC-0 256 Level     e6260000.mfis
15:          0      16971          0          0          0          0          0          0     GIC-0 257 Level     e6260000.mfis
16:          0          0        129          0          0          0          0          0     GIC-0 258 Level     e6260000.mfis
17:          0          0          0          1          0          0          0          0     GIC-0 259 Level     e6260000.mfis
18:          0          0          0          0          0          0          0          0     GIC-0 260 Level     e6260000.mfis
19:          0          0          0          0          0          0          0          0     GIC-0 261 Level     e6260000.mfis
20:          0          0          0          0          0          0          0          0     GIC-0 262 Level     e6260000.mfis
21:          0          0          0          0          0          0          0          0     GIC-0 263 Level     e6260000.mfis
---
 drivers/misc/rcar-mfis/rcar_mfis_drv.c | 5 +++++
 1 file changed, 5 insertions(+)

diff --git a/drivers/misc/rcar-mfis/rcar_mfis_drv.c b/drivers/misc/rcar-mfis/rcar_mfis_drv.c
index 48778719f2f2..959cf199ff23 100644
--- a/drivers/misc/rcar-mfis/rcar_mfis_drv.c
+++ b/drivers/misc/rcar-mfis/rcar_mfis_drv.c
@@ -158,6 +158,7 @@ static int rcar_mfis_probe(struct platform_device *pdev)
 	struct device *dev = &pdev->dev;
 	int i = 0;
 	u32 value = 0;
+	int cpu;
 
 	dev_dbg(dev, "R-Car MFIS probe start\n");
 
@@ -225,6 +226,10 @@ static int rcar_mfis_probe(struct platform_device *pdev)
 			continue;
 		}
 
+		/* set the initial affinity to reduce interrupt latency */
+		cpu = (i % num_online_cpus()) ? (i % num_online_cpus()) : 1;
+		irq_set_affinity_hint(irq->start, cpumask_of(cpu));
+
 		INIT_WORK(&mfis_ch->work, rcar_mfis_work);
 
 		mfis_ch->initialized = 1;
-- 
2.25.1

